pkg/identityserver/bunstore/user_bookmarks_store.go

```diff
@@ -0,0 +1,505 @@
+	case store.EntityApplication:
+		res.EntityIds = (&ttnpb.ApplicationIdentifiers{ApplicationId: m.EntityID}).GetEntityIdentifiers()
+	case store.EntityClient:
+		res.EntityIds = (&ttnpb.ClientIdentifiers{ClientId: m.EntityID}).GetEntityIdentifiers()
+	case store.EntityGateway:
+		res.EntityIds = (&ttnpb.GatewayIdentifiers{GatewayId: m.EntityID}).GetEntityIdentifiers()
+	case store.EntityOrganization:
+		res.EntityIds = (&ttnpb.OrganizationIdentifiers{OrganizationId: m.EntityID}).GetEntityIdentifiers()
+	case store.EntityUser:
+		res.EntityIds = (&ttnpb.UserIdentifiers{UserId: m.EntityID}).GetEntityIdentifiers()
+	case store.EntityEndDevice:
+		devIDs := strings.Split(m.EntityID, ".")
+		res.EntityIds = (&ttnpb.EndDeviceIdentifiers{
+			DeviceId:       devIDs[1],
+			ApplicationIds: &ttnpb.ApplicationIdentifiers{ApplicationId: devIDs[0]},
+		}).GetEntityIdentifiers()
+	default:
+		return nil, store.ErrInvalidEntityType.WithAttributes("entity_type", m.EntityType)
+	}
+
+	return res, nil
+}
+
+type userBookmarkStore struct{ *entityStore }
+
+func newUserBookmarkStore(baseStore *baseStore) *userBookmarkStore {
+	return &userBookmarkStore{entityStore: newEntityStore(baseStore)}
+}
+
+func (*userBookmarkStore) selectWithUserID(
+	_ context.Context, ids ...string,
+) func(*bun.SelectQuery) *bun.SelectQuery {
+	return func(q *bun.SelectQuery) *bun.SelectQuery {
+		switch len(ids) {
+		case 0:
+			return q
+		case 1:
+			return q.Where("?TableAlias.user_id = ?", ids[0])
+		default:
+			return q.Where("?TableAlias.user_id IN (?)", bun.In(ids))
+		}
+	}
+}
+
+func (*userBookmarkStore) selectWithEntityID(
+	ctx context.Context, ids ...ttnpb.IDStringer,
+) func(*bun.SelectQuery) *bun.SelectQuery {
+	return func(q *bun.SelectQuery) *bun.SelectQuery {
+		switch len(ids) {
+		case 0:
+			return q
+		case 1:
+			return q.Where(
+				"?TableAlias.entity_type = ? AND ?TableAlias.entity_id = ?",
+				ids[0].EntityType(),
+				ids[0].IDString(),
+			)
+		default:
+			entityByType := make(map[string][]string)
+			for _, id := range ids {
+				entityByType[id.EntityType()] = append(entityByType[id.EntityType()], id.IDString())
+			}
+			q = q.WhereGroup(" AND ", func(sq *bun.SelectQuery) *bun.SelectQuery {
+				for entityType, ids := range entityByType {
+					sq = sq.WhereOr(
+						"?TableAlias.entity_type = ? AND ?TableAlias.entity_id IN (?)",
+						entityType,
+						bun.In(ids),
+					)
+				}
+				return sq
+			})
+			return q
+		}
+	}
+}
+
+func (s *userBookmarkStore) getUserBookmarkBy(
+	ctx context.Context,
+	by func(*bun.SelectQuery) *bun.SelectQuery,
+	_ store.FieldMask,
+) (*UserBookmarks, error) {
+	model := &UserBookmarks{}
+	selectQuery := s.newSelectModel(ctx, model).Apply(by)
+
+	if err := selectQuery.Scan(ctx); err != nil {
+		return nil, storeutil.WrapDriverError(err)
+	}
+
+	return model, nil
+}
+
+func (s *userBookmarkStore) listUserBookmarksBy(
+	ctx context.Context,
+	by func(*bun.SelectQuery) *bun.SelectQuery,
+	_ store.FieldMask,
```

Comment:
Should we just remove the field mask from the API, given that we don't respect it in the list operation ? No matter what we do later on, for backwards compatibility purposes we will have to interpret the empty field mask as meaning at least the existing fields (`user_id`, `entity_id`, `created_at`).

Reply:
agree, was reluctant to remove it and choose to keep it as standard practice but indeed there is no need to have them, specially since I don't see bookmarks evolving past the point of being a reference between `<user_id>` and `(<entity_type>, <entity_id>)`